run_flax_speech_recognition_seq2seq_streaming_review.py

```diff
@@ -0,0 +1,1391 @@
+
+
+def create_learning_rate_fn(
+    num_train_steps: int, num_warmup_steps: int, learning_rate: float, start_step: int=0, warmup_init_value: float=0.0, decay_end_value: float=0.0,
+) -> Callable[[int], jnp.array]:
+    """Returns a linear warmup, linear_decay learning rate function."""
+    warmup_fn = optax.linear_schedule(
+        init_value=warmup_init_value, end_value=learning_rate, transition_steps=num_warmup_steps)
+    decay_fn = optax.linear_schedule(
+        init_value=learning_rate, end_value=decay_end_value, transition_steps=num_train_steps - num_warmup_steps
+    )
+    schedule_fn = optax.join_schedules(
+        schedules=[warmup_fn, decay_fn], boundaries=[num_warmup_steps])
+    
+    def learning_rate_fn(step: int) -> jnp.array:
+        return schedule_fn(step + start_step)
+    
+    return learning_rate_fn
+
+
+def main():
+    # Parse input arguments
+    # See all possible arguments in src/transformers/training_args.py
+    # or by passing the --help flag to this script.
+    # We now keep distinct sets of args, for a cleaner separation of concerns.
+    parser = HfArgumentParser(
+        (ModelArguments, DataTrainingArguments, Seq2SeqTrainingArguments))
+
+    if len(sys.argv) == 2 and sys.argv[1].endswith(".json"):
+        # If we pass only one argument to the script and it's the path to a json file,
+        # let's parse it to get our arguments.
+        model_args, data_args, training_args = parser.parse_json_file(
+            json_file=os.path.abspath(sys.argv[1]))
+    else:
+        model_args, data_args, training_args = parser.parse_args_into_dataclasses()
+
+    # Sending telemetry. Tracking the example usage helps us better allocate resources to maintain them. The
+    # information sent is the one passed as arguments along with your JAX/Flax versions.
+    send_example_telemetry("run_speech_recognition_seq2seq",
+                           model_args, data_args, framework="flax")
+
+    # Setup logging
+    # Make one log on every process with the configuration for debugging.
+    logging.basicConfig(
+        format="%(asctime)s - %(levelname)s - %(name)s - %(message)s",
+        datefmt="%m/%d/%Y %H:%M:%S",
+        handlers=[logging.StreamHandler(sys.stdout)],
+    )
+
+    # Number of hosts
+    num_of_hosts = jax.process_count()
+    current_host_idx = jax.process_index()
+
+    if current_host_idx == 0:
+        datasets.utils.logging.set_verbosity_warning()
+        transformers.utils.logging.set_verbosity_info()
+    else:
+        datasets.utils.logging.set_verbosity_error()
+        transformers.utils.logging.set_verbosity_error()
+    
+    logger.setLevel(logging.INFO)
+    logger.info("Training/evaluation parameters %s", training_args)
+
+    if num_of_hosts and not training_args.push_to_hub:
+        logger.warning(
+            f"If you are on a TPU Pod or a multinode setup, you need to set --push_to_hub to be able to save checkpoints to the hub."
```

Comment:
We'll rephrase it like this: ```If you are on a TPU Pod or a multinode setup, it is impossible to recover from local checkpoints. You will therefore need to set `--push_to_hub` to save checkpoints and be able to restore a training```

We're also fine removing the warning.

Reply:
Ah, is this because we only save on worker 0? https://github.com/NbAiLab/nb-whisper/blob/c066d3f69087dac21e125066fe25cc4eb48b7122/run_flax_speech_recognition_seq2seq_streaming.py#L1318-L1321

Maybe we can update this to save on all workers? That way we should be able to load a local checkpoint no?

Reply:
One argument of saving only on worker 0, and restoring only from the hub, is that we are certain that all the workers have loaded the same checkpoint. When we are restoring checkpoints like this, it is typically after a crash. One scenario for a crash is that one of the workers has a full disk, and has not been able to store the latest checkpoint. Complex errors might happen if we then restore the latest local checkpoint.